### PR TITLE
blockchain/vm: refactor precompiled contract codes and tests

### PR DESCRIPTION
**Pull request by yoomee1313**
## Proposed changes
- Before implementing klaytn's first incompatible change, code related to the precompiled contract is refactored.
- integrate RunVMLogContract, RunFeePayerContract, RunValidateSenderContract to RunPrecompiledContract
- refactor contracts_test.go and add testPrecompiledOOG test.
  - Most of them came from those commits: https://github.com/ethereum/go-ethereum/commit/4fc678542dd287c9759107d7f1f2291376434f5e, https://github.com/ethereum/go-ethereum/commit/7b189d6f1f7eedf46c6607901af291855b81112b.
  - take out test data from `contracts_test.go` to  `blockchain/vm/testdata/precompiles` folder
  - integrate precompile test functions to `testPrecompiled`
  - integrate precompile benchmark functions to `benchmarkPrecompiled`
- Test that precompiled contract is running exactly as before in two ways
  - Use [contracts_test.go](https://github.com/klaytn/klaytn/blob/dev/blockchain/vm/contracts_test.go). Well-written unit tests are located for all precompiled contracts.
  - Use [tests/precompiled_contracts_test.go](https://github.com/klaytn/klaytn/blob/dev/tests/precompiled_contracts_test.go) for call test of 0x01-0x08.
  - Use [tests/fee_payer_contract_test.go](https://github.com/klaytn/klaytn/blob/dev/tests/fee_payer_contract_test.go) for call test of 0x0a.
  - Use [tests/validate_sender_test.go](https://github.com/klaytn/klaytn/blob/dev/tests/validate_sender_test.go) for call test of 0x0b.
  - Use below contract for call test of 0x09. Run rpc call to setup VMLogTarget `> debug.setVMLogTarget(1)`
```
contract VmLogContract {
    bytes memory data = "Hello, Klaytn";
    uint256 idx = data.length;
    assembly {
        // skip length header.
        let ptr := add(data, 0x20)
        if iszero(call(100000, 0x09,0,ptr,idx,31,1)) {
            invalid()
        }
        return(0,32)
    }
    // same as below
    // address vmlogaddr = address(0x09);
    // vmlogaddr.call("Hello, Klaytn");
}
```
- Call Result of 0x09(vmLog). As we used VMLogTarget(1), this can be found in node's data folder(ex. data/ken/logs/vm.log)

dev

```
tx=0x0000000000000000000000000000000000000000000000000000000000000000 caller=0x044E43A75bB2e8A2d5b2328e05370715C377421D msg=Hello, Klaytn
tx=0x0000000000000000000000000000000000000000000000000000000000000000 caller=0x044E43A75bB2e8A2d5b2328e05370715C377421D msg=Hello, Klaytn
tx=0x51c954a6d2a3442d7e69c87c82a8b1721da18ea0f4bb3ac01da8a8ed646cd4dd caller=0x044E43A75bB2e8A2d5b2328e05370715C377421D msg=Hello, Klaytn
tx=0x51c954a6d2a3442d7e69c87c82a8b1721da18ea0f4bb3ac01da8a8ed646cd4dd caller=0x044E43A75bB2e8A2d5b2328e05370715C377421D msg=Hello, Klaytn
```
PR
```
tx=0x0000000000000000000000000000000000000000000000000000000000000000 caller=0x044E43A75bB2e8A2d5b2328e05370715C377421D msg=Hello, Klaytn
tx=0x0000000000000000000000000000000000000000000000000000000000000000 caller=0x044E43A75bB2e8A2d5b2328e05370715C377421D msg=Hello, Klaytn
tx=0x4b53393099e6d55ebe468040e071f867dd82a8edb3b4f1dcf8a93b662b6b5cd8 caller=0x044E43A75bB2e8A2d5b2328e05370715C377421D msg=Hello, Klaytn
tx=0x4b53393099e6d55ebe468040e071f867dd82a8edb3b4f1dcf8a93b662b6b5cd8 caller=0x044E43A75bB2e8A2d5b2328e05370715C377421D msg=Hello, Klaytn
```

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [x] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by KimKyungup** [comment]
Please take a look. @ehnuje @winnie-byun @aidan-kwon

**Review by ehnuje — DISMISSED** [review]

**Comment by ehnuje** [comment]
@KimKyungup  @winnie-byun @aidan-kwon
Dear reviewers, could you please review this change?

File: blockchain/vm/contracts.go [hunk 1]
```diff
@@ -56,40 +56,38 @@ type PrecompiledContract interface {
...
 // RunPrecompiledContract runs and evaluates the output of a precompiled contract.
-func RunPrecompiledContract(p PrecompiledContract, input []byte, contract *Contract) (ret []byte, computationCost uint64, err error) {
+func RunPrecompiledContract(p PrecompiledContract, input []byte, contract *Contract, evm *EVM) (ret []byte, computationCost uint64, err error) {
 	gas, computationCost := p.GetRequiredGasAndComputationCost(input)
 	if contract.UseGas(gas) {
-		ret, err := p.Run(input)
+		ret, err = p.Run(input, contract, evm)
 		return ret, computationCost, err
 	}
+	if p == (&validateSender{}) {
```
Comment (winnie-byun) [inline blockchain/vm/contracts.go:88]: This is always false because `&validateSender{}` always creates a new struct. Checkout below :
```
	a := &validateSender{}
	b := &validateSender{}
	t.Log(a == b) // false
```
How about implementing like this? : 
```
	if _, ok := p.(*validateSender); ok {
```

**Review by winnie-byun — COMMENTED** [review]
If this is from eth, could you add the link?

File: blockchain/vm/contracts_test.go [hunk 2]
```diff
@@ -21,783 +21,207 @@
...
 // precompiledTest defines the input/output pairs for precompiled contract tests.
 type precompiledTest struct {
-	input, expected string
-	gas             uint64
-	name            string
-	noBenchmark     bool // Benchmark primarily the worst-cases
+	Input, Expected string
+	Gas             uint64
+	Name            string
+	NoBenchmark     bool // Benchmark primarily the worst-cases
```
Comment (winnie-byun) [inline blockchain/vm/contracts_test.go:46]: Why did you change these names?

File: blockchain/vm/contracts.go [hunk 3]
```diff
@@ -448,51 +434,29 @@ func (c *feePayer) GetRequiredGasAndComputationCost(input []byte) (uint64, uint6
...
-		return []byte{0}, computationCost, nil
+		return []byte{0}, nil
 	}
-
-	return []byte{1}, computationCost, nil
+	return []byte{1}, nil
 }
 
-func validateSender(input []byte, picker types.AccountKeyPicker) error {
+func validate(input []byte, picker types.AccountKeyPicker) error {
```
Comment (aidan-kwon) [inline blockchain/vm/contracts.go:459]: Can I know why did you change this function name? I think 'validate' is too general and hard to find the connection with `validateSender` struct. How about using the original name or converting this function to a method of `validateSender` like `(c *validateSender) do`.

File: blockchain/vm/contracts.go [hunk 4]
```diff
@@ -56,40 +56,38 @@ type PrecompiledContract interface {
...
-func RunPrecompiledContract(p PrecompiledContract, input []byte, contract *Contract) (ret []byte, computationCost uint64, err error) {
+func RunPrecompiledContract(p PrecompiledContract, input []byte, contract *Contract, evm *EVM) (ret []byte, computationCost uint64, err error) {
 	gas, computationCost := p.GetRequiredGasAndComputationCost(input)
 	if contract.UseGas(gas) {
-		ret, err := p.Run(input)
+		ret, err = p.Run(input, contract, evm)
 		return ret, computationCost, err
 	}
+	if p == (&validateSender{}) {
+		return []byte{0}, computationCost, kerrors.ErrOutOfGas
```
Comment (KimKyungup) [inline blockchain/vm/contracts.go:89]: This additional returning logic is necessary?
How about `p.Run()  returning this?

**Comment by KimKyungup** [comment]
@yoomee1313 Is there any progress?

File: blockchain/vm/contracts.go (hunk 4 above)
Comment (yoomee1313) [reply blockchain/vm/contracts.go:89]: Sorry for the late, I checked whether this logic is necessary or not. 
To tell you the conclusion, it is unnecessary logic. The above logic occurs when out-of-gas occurs. That is, it happens when internal tx is called with insufficient gas. Internal tx is made by `assembly call`.
According to the Solidity documentation, the `assembly call` returns `0` as a return value when an error such as out-of-gas occurs. Therefore, even if only kerrors.ErrOutOfGas is returned in the code above, the return value is automatically returned as 0.

File: blockchain/vm/contracts.go (hunk 1 above)
Comment (yoomee1313) [reply blockchain/vm/contracts.go:88]: Thanks, it worked well!! However, according to below comment, i rechecked whether this logic is necessary or not. In conclusion, it is not necessary, so it will be deleted!

File: blockchain/vm/contracts.go (hunk 3 above)
Comment (yoomee1313) [reply blockchain/vm/contracts.go]: I reflected your review at a96ffc7612723b72b3b3c9df756bb7bc3ddcfc27! Thank you.

File: blockchain/vm/contracts_test.go (hunk 2 above)
Comment (yoomee1313) [reply blockchain/vm/contracts_test.go:46]: I have to capitalize them to import test datas from json file.. Umm, do you have any idea..?

Comment (winnie-byun) [reply blockchain/vm/contracts_test.go:46]: It just looked strange because the struct name is not capitalized and the fields are capitalized.
It looks fine to me :)

**Comment by yoomee1313** [comment]
@KimKyungup @winnie-byun @aidan-kwon I reflected your reviews. The change is https://github.com/klaytn/klaytn/commit/a96ffc7612723b72b3b3c9df756bb7bc3ddcfc27.

File: blockchain/vm/interpreter.go [hunk 5]
```diff
@@ -82,13 +75,15 @@ type Interpreter struct {
...
+// NewEVMInterpreter returns a new instance of the Interpreter.
+func NewEVMInterpreter(evm *EVM, cfg *Config) *Interpreter {
 	// We use the STOP instruction whether to see
 	// the jump table was initialised. If it was not
 	// we'll set the default jump table.
 	if !cfg.JumpTable[STOP].valid {
-		cfg.JumpTable = ConstantinopleInstructionSet
+		var jt JumpTable
+		jt = ConstantinopleInstructionSet
+		cfg.JumpTable = jt
```
Comment (ehnuje) [inline blockchain/vm/interpreter.go:86]: Why do we need this change instead of direct assignment like before?

**Review by winnie-byun — APPROVED** [review]

File: blockchain/vm/interpreter.go (hunk 5 above)
Comment (yoomee1313) [reply blockchain/vm/interpreter.go:86]: thanks, i'll remove it.

Comment (yoomee1313) [inline blockchain/vm/interpreter.go:86]:
```suggestion
		cfg.JumpTable = ConstantinopleInstructionSet
```

File: blockchain/vm/contracts.go [hunk 6]
```diff
@@ -456,7 +453,7 @@ func (c *validateSender) Run(input []byte, contract *Contract, evm *EVM) ([]byte
 	return []byte{1}, nil
```
Comment (KimKyungup) [inline blockchain/vm/contracts.go:453]: Why doesn't this return the error? @yoomee1313

Comment (yoomee1313) [reply blockchain/vm/contracts.go:453]: @KimKyungup `vmLog` is designed to return 0 if sender validation is failed, instead returning error. It return 1 when validation is succeed.

**Review by aidan-kwon — DISMISSED** [review]
LGTM

**Review by ehnuje — APPROVED** [review]

**Review by KimKyungup — DISMISSED** [review]

**Comment by yoomee1313** [comment]
Updated description

**Comment by yoomee1313** [comment]
Resolved conflict https://github.com/klaytn/klaytn/pull/869/commits/d523a9c16edb56f013e52dbca98ec2b9d3abbb0c :)

**Review by ehnuje — APPROVED** [review]

**Review by aidan-kwon — APPROVED** [review]

**Review by KimKyungup — APPROVED** [review]

**Review by winnie-byun — APPROVED** [review]
